### PR TITLE
Updating ext_proc filter clear route cache logic.

### PR DESCRIPTION
**Pull request by jstraceski**
Commit Message: Modifying clear route cache logic for the ext_proc filter.
Additional Description: Adding in the ability to disable route cache clearing ont he ext_proc filter. Changing the behavior of the clearing to check if there is a header mutation.
Risk Level: Low
Testing: [test/extensions/filters/http/ext_proc/filter_test.cc](test/extensions/filters/http/ext_proc/filter_test.cc)
Docs Changes: [docs/root/configuration/http/http_filters/ext_proc_filter.rst](docs/root/configuration/http/http_filters/ext_proc_filter.rst)
Release Notes: [changelogs/current.yaml](changelogs/current.yaml)
Platform Specific Features:

### CODE REVIEW

File: api/envoy/config/common/mutation_rules/v3/mutation_rules.proto
Comment (htuch) [inline api/envoy/config/common/mutation_rules/v3/mutation_rules.proto:91]: This isn't a header mutation rule. We probably need some other top-level attribute restriction.

Comment (jstraceski) [reply api/envoy/config/common/mutation_rules/v3/mutation_rules.proto]: Done.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (rshriram) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:175]: please add another comment that this is true by default. (for backward compat purposes)

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (rshriram) [inline source/extensions/filters/http/ext_proc/processor_state.cc:96]: This should be true by default

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (rshriram) [inline source/extensions/filters/http/ext_proc/processor_state.cc:330]: true by default

Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (jstraceski) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:175]: Done.

File: source/extensions/filters/common/mutation_rules/mutation_rules.h
Comment (tyxia) [inline source/extensions/filters/common/mutation_rules/mutation_rules.h:44]: This function seems not be used. Remove?

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:95]: This flag should be available when this class is constructed and its value is consistent.

Can we add a member variable for this and only init it in constructor? so that we can avoid retrieving it every time when this function is called, which is redundant

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:329]: ditto

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:97]: nit: The comment here needs to be updated.  Because in addition to the mutation to the header, we are also checking this new `allow_clearing_route_cache`

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:99]: The change of request header could affect the route selection so we need to clear route cache to trigger the route re-selection (so that we don't read from cache directly). The body mutation will not affect.
Can you remove has_body_mutation check?

Also, one nit. Can you move `allow_clearing_route_cache` to the first check. Slightly better for performance since we will not check the other conditions if first one failed.

**Review by tyxia — COMMENTED** [review]
Thanks for working on this! Left some comments.
Besides:
- Could you run the command `sudo ./ci/run_envoy_docker.sh '\''./ci/do_ci.sh format'\'' && sh bazel/setup_clang.sh /usr' `  to try to fix format, so that we can kickoff CI/CD
- Could you add a simple unit test?  You probably can refer to `ClearRouteCache` test in ext_proc/filter_test.cc

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Moved it to the filter config. Followed the convention there.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Removed body check and changed order.

File: source/extensions/filters/common/mutation_rules/mutation_rules.h
Comment (jstraceski) [reply source/extensions/filters/common/mutation_rules/mutation_rules.h]: Done.

**Review by tyxia — COMMENTED** [review]
We are getting closer.  I left some comments about test and code comment

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (jstraceski) [reply test/extensions/filters/http/ext_proc/filter_test.cc]: I had both of those tests to show that body modification doesn't trigger the clearing.
ClearRouteCacheUnchanged does no header or body modification.
ClearRouteCacheHeaderMutation does only header modification.
ClearRouteCacheBodyMutation does only body modification.
ClearRouteCacheDiasallowed does header modification while disabling the cache clearing (fixed this).

I could remove the body modification test because it doesn't check any extra code paths from the unchanged one.

With the new changes I have:
ClearRouteCacheUnchanged: does no header or body modification, expecting no cache clearing callbacks.
ClearRouteCacheHeaderMutation: does only header modification, expecting cache clearing callbacks..
ClearRouteCacheDiasallowedHeaderMutation: does header modification while disabling the cache clearing, expecting no cache clearing callbacks.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (jstraceski) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:174]: Done.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Done.

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (adisuissa) [inline api/envoy/service/ext_proc/v3/external_processor.proto:292]: Is the link a self-referral (referring to the same field)?
Perhaps it should be the field added in the ext_proc filter above?

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:173]: I think the description can be clarified a bit:
```
Prevents clearing the route-cache when the clear_route_cache<link> field is set in an external processor response. If true, the response field is ignored. By default, this field is set to false, to ensure backwards compatibility.
```

**Review by adisuissa — COMMENTED** [review]
Thanks for adding this!
Left minor comments on the API.

A high-level comment (probably not in the scope of this PR), from the point-of-view of the external processor, it will be difficult to reason why the external-processor sent a response to clear the route cache, but the Envoy did not act accordingly. Specifically, the provider of an ext-proc service will see that their service works for Envoy deployment A (that allows cache clearing) but not for Envoy deployment B (that prevents cache clearing), but will not be able to debug why this is happening.
A better approach may be to "publish a contract" when connecting to the external-processor stating what features are enabled/disabled. @htuch WDYT?

**Comment by tyxia** [comment]
> Thanks for adding this! Left minor comments on the API.
> 
> A high-level comment (probably not in the scope of this PR), from the point-of-view of the external processor, it will be difficult to reason why the external-processor sent a response to clear the route cache, but the Envoy did not act accordingly. Specifically, the provider of an ext-proc service will see that their service works for Envoy deployment A (that allows cache clearing) but not for Envoy deployment B (that prevents cache clearing), but will not be able to debug why this is happening. A better approach may be to "publish a contract" when connecting to the external-processor stating what features are enabled/disabled. @htuch WDYT?

The context/intention (we probably could add it to PR description) is:  third-party extension server (which send back response) can do whatever they want. On the other hand, our product (which create ext_proc filter through filter config) want to force clearing behavior off regardless extension's behavior (e.g., due to product requirement, security, etc). So, ext_proc don't really need to set up the contract with extension or report for permission.

I agree it is good to improve the observability/debuggability. We can add the log (,even stats) when `common_response.clear_route_cache() == true && filter_.config().disableRouteCacheClearing() == true` for this behavior. WDYT?

**Comment by adisuissa** [comment]
> The context/intention (we probably could add it to PR description) is: third-party extension server (which send back response) can do whatever they want. On the other hand, our product (which create ext_proc filter through filter config) want to force clearing behavior off regardless extension's behavior (e.g., due to product requirement, security, etc). So, ext_proc don't really need to set up the contract with extension or report for permission.
> 
I agree that from a security perspective this is a valid feature. FWIW, other fields should also be restricted, depending on the threat-model.

I think we should address the general case, where the extension and the deployment are handled by different entities, and the extension can either be a third-part extensions server or a specific service. The case of an extensions server is probably easier (if the extension server and Envoy are managed by the same entity), as the extensions server should prevent accepting extensions that clear the routes, and this "contract" should be enforced by the extensions server.
The added benefits for a "contract" for the extensions server are: 1) it can clearly reason what kind of extensions can be executed using the Envoy; 2) allows the operators to debug things on their end and do not need to access Envoy's logs/stats. I think it is better to decouple the things as much as possible, and as things stand, if the Envoy deployer updates the ext-proc config on their side, they may break the third-party extension server.

> I agree it is good to improve the observability/debuggability. We can add the log (,even stats) when `common_response.clear_route_cache() == true && filter_.config().disableRouteCacheClearing() == true` for this behavior. WDYT?

I thought of suggesting counters (IMHO logs are typically less useful in large deployments), but the downsides are that stats require more memory, are useful to know if an event happens but not so useful to understand what triggered that event (e.g., which response from the external processor caused it), and don't scale if a similar fix will be needed on different fields.

**Comment by tyxia** [comment]
> > The context/intention (we probably could add it to PR description) is: third-party extension server (which send back response) can do whatever they want. On the other hand, our product (which create ext_proc filter through filter config) want to force clearing behavior off regardless extension's behavior (e.g., due to product requirement, security, etc). So, ext_proc don't really need to set up the contract with extension or report for permission.
> 
> I agree that from a security perspective this is a valid feature. FWIW, other fields should also be restricted, depending on the threat-model.
> 
> I think we should address the general case, where the extension and the deployment are handled by different entities, and the extension can either be a third-part extensions server or a specific service. The case of an extensions server is probably easier (if the extension server and Envoy are managed by the same entity), as the extensions server should prevent accepting extensions that clear the routes, and this "contract" should be enforced by the extensions server. The added benefits for a "contract" for the extensions server are: 1) it can clearly reason what kind of extensions can be executed using the Envoy; 2) allows the operators to debug things on their end and do not need to access Envoy's logs/stats. I think it is better to decouple the things as much as possible, and as things stand, if the Envoy deployer updates the ext-proc config on their side, they may break the third-party extension server.
> 
Please first let me confirm I understand your "contract" and "this contract should be enforced by the extensions server" correctly. Are you suggesting : 1) ext_proc reports that setting `clear_route_cache` is off to the Server (to which ext_proc call out) 2) Sever evaluate it and make the final decision.  
AFAIU, this is contrary to our goal. what we want to achieve because the server can choose to ignore ext_proc setting and that is out of our control. 
I think ext_proc can report its setting/feature to server to help their observability but our end goal here is enforcement on ext_proc side not server side. @htuch Please correct me if I misunderstand the motivation behind this change.

**Comment by adisuissa** [comment]
> Please first let me confirm I understand your "contract" and "this contract should be enforced by the extensions server" correctly. Are you suggesting : 1) ext_proc reports that setting `clear_route_cache` is off to the external processor (to which ext_proc call out) 2) External processor evaluate it and make the final decision whether clear route cache or not.
> 
> AFAIU, this is contrary to what we want to achieve here because the external processor can choose to ignore ext_proc setting and that is out of our control.
> 
> I think ext_proc can report its setting/feature to external processor to help their observability but our end goal here is enforcement on ext_proc side not external processor side. @htuch Please correct me if I misunderstand our goal.

This is similar to how Envoy talks with an xDS-server and sends a list of all supported extensions. It allows the server to decide what kind of configs to send back. If the server sends a config with a wrong extension, the client can reject the config.

In the ext-proc case, the Envoy will send the "contract" (supported abilities). If it is configured to disable route-cleaning, it will disregard whatever the ext-proc service returns in the field (just as this PR suggests). The only difference is that the ext-proc service operator can clearly observe that the Envoy that connects to it does or doesn't support some feature and they can debug things on their end without needing to look at the logs/counters on the Envoy side.

**Comment by tyxia** [comment]
> > Please first let me confirm I understand your "contract" and "this contract should be enforced by the extensions server" correctly. Are you suggesting : 1) ext_proc reports that setting `clear_route_cache` is off to the external processor (to which ext_proc call out) 2) External processor evaluate it and make the final decision whether clear route cache or not.
> > AFAIU, this is contrary to what we want to achieve here because the external processor can choose to ignore ext_proc setting and that is out of our control.
> > I think ext_proc can report its setting/feature to external processor to help their observability but our end goal here is enforcement on ext_proc side not external processor side. @htuch Please correct me if I misunderstand our goal.
> 
> This is similar to how Envoy talks with an xDS-server and sends a list of all supported extensions. It allows the server to decide what kind of configs to send back. If the server sends a config with a wrong extension, the client can reject the config.
> 
> In the ext-proc case, the Envoy will send the "contract" (supported abilities). If it is configured to disable route-cleaning, it will disregard whatever the ext-proc service returns in the field (just as this PR suggests). The only difference is that the ext-proc service operator can clearly observe that the Envoy that connects to it does or doesn't support some feature and they can debug things on their end without needing to look at the logs/counters on the Envoy side.

I am not sure if we need full/similar capability of xDS for service callout since data <->control plane seems to be more complicated (at least at this moment). Though, the mechanism you described sound like a good design . 

> If it is configured to disable route-cleaning, it will disregard whatever the ext-proc service returns in the field (just as this PR suggests)
Since we need this discard behavior anyways to enforce ext_proc side enforcement, then IMHO we can proceed with this PR.  Adding contracts can be a separate part that will be added later (e.g., as post-MVP) if it is determined to be desired for service callout. WDYT?

**Comment by adisuissa** [comment]
> Since we need this discard behavior anyways to enable ext_proc side enforcement, then IMHO we can proceed with this PR. Adding contracts can be a separate part that will be added later (e.g., as post-MVP) if it is determined to be desired for service callout. WDYT?

Yeah, as stated in the first comment, this is probably not part of this PR.

**Comment by htuch** [comment]
@adisuissa yeah, I can see this being something to add in the future as the range of behaviors becomes harder to debug.

I do think we need something here that is more than silent failure though, that is basically impossible to debug, even in the situation where you control client / server. Typically a stat would make sense and/or we could consider this a protocol violation and terminate the connection with the ext_proc server.

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (jstraceski) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: Ah, yep. Put it in the wrong location.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (jstraceski) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto]: Done.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (jstraceski) [reply test/extensions/filters/http/ext_proc/filter_test.cc:2013]: Addressed this in the previous comment.

**Review by tyxia — COMMENTED** [review]
Mostly LGTM, Could you add the stat to track the case that  `common_response.clear_route_cache() = true && filter_.config().disableRouteCacheClearing() = true `, as discussed in the comments.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:175]: Please add a reference to this field in the comment of `CommonResponse.clear_route_cache`.
Also, WDYT about renaming the field to `disable_clear_route_cache`? I don't feel strongly about the name, but I think the motivation is when searching for a name in the docs, it should be able to return results for the two fields.

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left minor comments on the API.
Also +1 on the prior [comment](https://github.com/envoyproxy/envoy/pull/26388#pullrequestreview-1389196688) to add a stat to track this.
/wait

**Review by tyxia — APPROVED** [review]
Thanks!

LGTM, a few small nits.

**Comment by jstraceski** [comment]
I was having some client side issues with the status of the branch.
I believe that I initially branched it from something that wasn't main.
It should be in sync now though, I'm unsure why it keeps getting out of sync.
I have tried `git pull upstream main` but that also complains about ff, and `git pull --rebase upstream main` makes it complain when I push. I believe that it should be fixed now.

**Review by adisuissa — APPROVED** [review]
/lgtm api
Assigning @snowp as a non-Googler reviewer and code-owner.
/assign @snowp

**Comment by adisuissa** [comment]
Please also check code format (see CI errors).

**Comment by adisuissa** [comment]
CI error:
`/tmp/tmp2etuagir/generated/rst/version_history/v1.27/v1.27.0.rst:76: WARNING: undefined label: 'envoy_v3_api_field_service.ext_proc.v3.externalprocessor.disable_clear_route_cache'`

seems that the field identifier is incorrect. Can you PTAL?

**Comment by adisuissa** [comment]
Snow is OOO.
@wbpcode can you review this code as a non-Googler?
/assign @wbpcode

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (wbpcode) [inline source/extensions/filters/http/ext_proc/processor_state.h:109]: `const envoy::service::ext_proc::v3::CommonResponse&` to avoid unnecessary copy.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (wbpcode) [inline test/extensions/filters/http/ext_proc/filter_test.cc:2026]: didn't get why? for some code format unification?

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (wbpcode) [inline test/extensions/filters/http/ext_proc/filter_test.cc:2039]: Seems that the test name is not match the logic.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contributions. And only some comments need to be addressed, others LGTM.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (jstraceski) [reply test/extensions/filters/http/ext_proc/filter_test.cc:2026]: The ordering of the checks was modified to be clearer to read.
https://en.wikipedia.org/wiki/Yoda_conditions

Comment (tyxia) [reply test/extensions/filters/http/ext_proc/filter_test.cc:2026]: There is a c++ readability advice called `Avoid "Yoda-style" comparisons`  at Google. I didn't find its open-sourced version like other abseil tips of week. I just copied&pasted here for reference:
```
In English, which of the following sounds more natural:

"If var is equal to 42"
"If 42 is equal to var"
Likewise, in code, prefer var == 42 to 42 == var. More generally, put the "most variable" operand on the left 
(so, for example, when comparing a function local variable to a class member variable, one would usually put the local variable on the left).
```

It's not one of Envoy's [deviation](https://github.com/envoyproxy/envoy/blob/main/STYLE.md#deviations-from-google-c-style-guidelines) and I think it makes sense for new code/file follow this rule as it is more natural

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.h:108]: It could be a private function? But let's keep it as it is since there are several other functions like `hasBufferedData`. We could clean them up in a separate PR if needed.

**Review by tyxia — APPROVED** [review]
Thanks!

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
ping @adisuissa to merge after ci is done.

**Review by tyxia — DISMISSED** [review]

**Comment by adisuissa** [comment]
Needs a senior maintainer review.
/assign @htuch

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:378]: LGTM, but can I ask why we bundled the "avoid clearing route cache when no header mutations" into this PR?

IMHO it probably makes sense, although there might be valid reasons (albeit contrived) to use ext_proc to force a repick without this, e.g. if an earlier filter had set some headers and we're not having them take effect. As I said, a bit contrived and it's probably fine, just not sure why we needed it in this PR.

Comment (jstraceski) [reply source/extensions/filters/http/ext_proc/processor_state.cc:378]: I had included it in the scope of this pr earlier on because I had assumed it was part of the plan for the change, it was mentioned in one of the readiness documents in some notes. It seemed related enough to include. 
Though, it does make sense to split it up into another pr. What do you think?

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/processor_state.cc:378]: IMHO, if earlier filter mutate the header to affect route selection, technically that filter should clear route cache itself in place. Leaving it to other filters to clear seems not very robust behavior to me.

Yea, i don't have strong opinion. Either this PR or separate PR works for me.

Comment (htuch) [reply source/extensions/filters/http/ext_proc/processor_state.cc:378]: OK, I don't see any objections to combining, but please keep in mind for future PRs that ideally we have orthogonal PRs to support things like bisecting and change / risk management.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
